File: FilFeed/src/main/java/com/phicdy/filfeed/ui/SettingFragment.java
package com.phicdy.filfeed.ui;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.SwitchPreference;
import android.widget.Toast;

import com.phicdy.filfeed.BuildConfig;
import com.phicdy.filfeed.R;
import com.phicdy.filfeed.alarm.AlarmManagerTaskManager;
import com.phicdy.filfeed.db.DatabaseAdapter;
import com.phicdy.filfeed.util.PreferenceHelper;
import com.phicdy.filfeed.util.ToastHelper;

public class SettingFragment extends PreferenceFragment {

    private ListPreference prefUpdateInterval;
    private ListPreference prefAllReadBehavior;
    private ListPreference prefSwipeDirection;
    private SwitchPreference prefArticleSort;
    private SwitchPreference prefInternalBrowser;

    private PreferenceHelper helper;
    private SharedPreferences.OnSharedPreferenceChangeListener listener;

    public SettingFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (BuildConfig.DEBUG) {
            addPreferencesFromResource(R.xml.setting_fragment_debug);
        } else {
            addPreferencesFromResource(R.xml.setting_fragment);
        }
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        helper = PreferenceHelper.getInstance(getActivity());
        initView();
        initListener();
    }

    @Override
    public void onResume() {
        super.onResume();
        getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(listener);
    }

    @Override
    public void onPause() {
        super.onPause();
        getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(listener);
    }

    private void initView() {
        // Set defalut value of update interval
        prefUpdateInterval = (ListPreference)findPreference(getString(R.string.key_update_interval));
        // Calc interval hour from saved interval second
        int autoUpdateIntervalSecond = helper.getAutoUpdateIntervalSecond();
        int autoUpdateIntervalHour = autoUpdateIntervalSecond / (60 * 60);
        String updateIntervalHourItems[] = getResources().getStringArray(R.array.update_interval_items_values);
        // Set index of saved interval
        for (int i = 0; i < updateIntervalHourItems.length; i++) {
            if (Integer.valueOf(updateIntervalHourItems[i]) == autoUpdateIntervalHour) {
                prefUpdateInterval.setValueIndex(i);
                // Set summary of saved interval
                String updateIntervalStringItems[] = getResources().getStringArray(R.array.update_interval_items);
                prefUpdateInterval.setSummary(updateIntervalStringItems[i]);
                break;
            }
        }

        // Set default value of article sort option
        prefArticleSort = (SwitchPreference)findPreference(getString(R.string.key_article_sort));
        prefArticleSort.setChecked(helper.getSortNewArticleTop());

        // Set default value of internal browser option
        prefInternalBrowser = (SwitchPreference)findPreference(getString(R.string.key_internal_browser));
        prefInternalBrowser.setChecked(helper.isOpenInternal());

        prefAllReadBehavior = (ListPreference)findPreference(getString(R.string.key_all_read_behavior));
        String allReadBehaviorItems[] = getResources().getStringArray(R.array.all_read_behavior_values);
        // Set index of behavior of all read
        for (int i = 0; i < allReadBehaviorItems.length; i++) {
            boolean allBehaviorItemBool = (Integer.valueOf(allReadBehaviorItems[i]) == 1);
            boolean savedValue = helper.getAllReadBack();
            if (allBehaviorItemBool == savedValue) {
                prefAllReadBehavior.setValueIndex(i);
                // Set summary of behavior of all read
                String allReadBehaviorStringItems[] = getResources().getStringArray(R.array.all_read_behavior);
                prefAllReadBehavior.setSummary(allReadBehaviorStringItems[i]);
                break;
            }
        }

        prefSwipeDirection = (ListPreference)findPreference(getString(R.string.key_swipe_direction));
        String swipeDirectionItems[] = getResources().getStringArray(R.array.swipe_direction_items_values);
        // Set index of swipe direction
        for (int i = 0; i < swipeDirectionItems.length; i++) {
            if (Integer.valueOf(swipeDirectionItems[i]) == helper.getSwipeDirection()) {
                prefSwipeDirection.setValueIndex(i);
                // Set summary of swipe direction
                String swipeDirectionStringItems[] = getResources().getStringArray(R.array.swipe_direction_items);
                prefSwipeDirection.setSummary(swipeDirectionStringItems[i]);
                break;
            }
        }
    }

    private void initListener() {
        listener = new SharedPreferences.OnSharedPreferenceChangeListener() {
            @Override
            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
                if (key.equals(getString(R.string.key_update_interval))) {
                    updateUpdateInterval();
                } else if (key.equals(getString(R.string.key_all_read_behavior))) {
                    updateAllReadBehavior();
                } else if (key.equals(getString(R.string.key_swipe_direction))) {
                    updateSwipeDirection();
                } else if (key.equals(getString(R.string.key_article_sort))) {
                    updateArticleSort();
                } else if (key.equals(getString(R.string.key_internal_browser))) {
                    updateInternalBrowser();
                }
            }
        };
        if (BuildConfig.DEBUG) {
            Preference prefImport = findPreference(getString(R.string.key_import_db));
            prefImport.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    DatabaseAdapter.getInstance(getActivity()).importDB();
                    ToastHelper.showToast(getActivity(), getString(R.string.import_db), Toast.LENGTH_SHORT);
                    return true;
                }
            });
            Preference prefExport = findPreference(getString(R.string.key_export_db));
            prefExport.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    DatabaseAdapter.getInstance(getActivity()).exportDb();
                    ToastHelper.showToast(getActivity(), getString(R.string.export_db), Toast.LENGTH_SHORT);
                    return true;
                }
            });
        }
    }

    public void updateUpdateInterval() {
        // Save new interval second
        int intervalHour = Integer.valueOf(prefUpdateInterval.getValue());
        int intervalSecond = intervalHour * 60 * 60;
        helper.setAutoUpdateIntervalSecond(intervalSecond);
        String updateIntervalHourItems[] = getResources().getStringArray(R.array.update_interval_items_values);

        // Refresh summary
        for (int i = 0; i < updateIntervalHourItems.length; i++) {
            if (Integer.valueOf(updateIntervalHourItems[i]) == intervalHour) {
                String updateIntervalStringItems[] = getResources().getStringArray(R.array.update_interval_items);
                prefUpdateInterval.setSummary(updateIntervalStringItems[i]);
                break;
            }
        }

        // Set new alarm
        AlarmManagerTaskManager.setNewAlarm(getActivity());
    }

    public void updateAllReadBehavior() {
        // Save new behavior of all read
        boolean isAllReadBack = (Integer.valueOf(prefAllReadBehavior.getValue()) == 1);
        helper.setAllReadBack(isAllReadBack);
        String allReadBehaviorItems[] = getResources().getStringArray(R.array.all_read_behavior_values);

        // Refresh summary
        for (int i = 0; i < allReadBehaviorItems.length; i++) {
            boolean allBehaviorItemBool = (Integer.valueOf(allReadBehaviorItems[i]) == 1);
            if (allBehaviorItemBool == isAllReadBack) {
                String allReadBehaviorStringItems[] = getResources().getStringArray(R.array.all_read_behavior);
                prefAllReadBehavior.setSummary(allReadBehaviorStringItems[i]);
                break;
            }
        }
    }

    public void updateSwipeDirection() {
        // Save new swipe direction
        int swipeDirection = Integer.valueOf(prefSwipeDirection.getValue());
        helper.setSwipeDirection(swipeDirection);
        String swipeDirectionItems[] = getResources().getStringArray(R.array.swipe_direction_items_values);

        // Refresh summary
        for (int i = 0; i < swipeDirectionItems.length; i++) {
            if (Integer.valueOf(swipeDirectionItems[i]) == swipeDirection) {
                String swipeDirectionStringItems[] = getResources().getStringArray(R.array.swipe_direction_items);
                prefSwipeDirection.setSummary(swipeDirectionStringItems[i]);
                break;
            }
        }
    }

    public void updateArticleSort() {
        helper.setSortNewArticleTop(prefArticleSort.isChecked());
    }

    public void updateInternalBrowser() {
        helper.setOpenInternal(prefInternalBrowser.isChecked());
    }

}

File: mycuration/src/main/java/com/phicdy/mycuration/presentation/view/fragment/SettingFragment.kt
package com.phicdy.mycuration.presentation.view.fragment

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreference
import com.phicdy.mycuration.BuildConfig
import com.phicdy.mycuration.R
import com.phicdy.mycuration.data.db.DatabaseHelper
import com.phicdy.mycuration.domain.alarm.AlarmManagerTaskManager
import com.phicdy.mycuration.presentation.presenter.SettingPresenter
import com.phicdy.mycuration.presentation.view.SettingView
import com.phicdy.mycuration.presentation.view.activity.LicenseActivity
import com.phicdy.mycuration.presentation.view.activity.UserRequestActivity
import com.phicdy.mycuration.tracker.TrackerHelper
import com.phicdy.mycuration.util.ToastHelper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.koin.android.scope.ext.android.bindScope
import org.koin.android.scope.ext.android.getOrCreateScope
import org.koin.core.parameter.parametersOf
import kotlin.coroutines.CoroutineContext


class SettingFragment : PreferenceFragmentCompat(), SettingView, CoroutineScope {

    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    private val presenter: SettingPresenter by inject { parametersOf(this) }

    private lateinit var prefUpdateInterval: ListPreference
    private lateinit var prefLaunchTab: ListPreference
    private lateinit var prefAllReadBehavior: ListPreference
    private lateinit var prefSwipeDirection: ListPreference
    private lateinit var prefAutoUpdateInMainUi: SwitchPreference
    private lateinit var prefArticleSort: SwitchPreference
    private lateinit var prefInternalBrowser: SwitchPreference
    private lateinit var prefLicense: Preference
    private lateinit var prefReview: Preference
    private lateinit var prefRequest: Preference

    private var listener: SharedPreferences.OnSharedPreferenceChangeListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bindScope(getOrCreateScope("setting"))
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        if (BuildConfig.DEBUG) {
            addPreferencesFromResource(R.xml.setting_fragment_debug)
        } else {
            addPreferencesFromResource(R.xml.setting_fragment)
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        presenter.activityCreate()
    }

    override fun onResume() {
        super.onResume()
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
    }

    override fun onPause() {
        super.onPause()
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    override fun initView() {
        prefUpdateInterval = findPreference(getString(R.string.key_update_interval)) as ListPreference
        prefAutoUpdateInMainUi = findPreference(getString(R.string.key_auto_update_in_main_ui)) as SwitchPreference
        prefArticleSort = findPreference(getString(R.string.key_article_sort)) as SwitchPreference
        prefInternalBrowser = findPreference(getString(R.string.key_internal_browser)) as SwitchPreference
        prefAllReadBehavior = findPreference(getString(R.string.key_all_read_behavior)) as ListPreference
        prefSwipeDirection = findPreference(getString(R.string.key_swipe_direction)) as ListPreference
        prefLaunchTab = findPreference(getString(R.string.key_launch_tab)) as ListPreference
        prefLicense = findPreference(getString(R.string.key_license))
        prefReview = findPreference(getString(R.string.key_review))
        prefRequest = findPreference(getString(R.string.key_request))
    }

    override fun initListener() {
        listener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
            when (key) {
                getString(R.string.key_update_interval) -> {
                    val intervalHour = Integer.valueOf(prefUpdateInterval.value)
                    val manager = AlarmManagerTaskManager(activity as Context)
                    presenter.updateUpdateInterval(intervalHour, manager)
                    // GA
                    TrackerHelper.sendSettingEvent(getString(R.string.change_auto_update_interval), intervalHour.toLong().toString())
                }
                getString(R.string.key_all_read_behavior) -> {
                    val isAllReadBack = Integer.valueOf(prefAllReadBehavior.value) == 1
                    presenter.updateAllReadBehavior(isAllReadBack)
                    // GA
                    val allReadBehaviorStringItems = resources.getStringArray(R.array.all_read_behavior)
                    TrackerHelper.sendSettingEvent(getString(R.string.change_all_read_behavior),
                            allReadBehaviorStringItems[Integer.valueOf(prefAllReadBehavior.value)])
                }
                getString(R.string.key_swipe_direction) -> {
                    val swipeDirection = Integer.valueOf(prefSwipeDirection.value)
                    presenter.updateSwipeDirection(swipeDirection)
                    // GA
                    val swipeDirectionStringItems = resources.getStringArray(R.array.swipe_direction_items)
                    TrackerHelper.sendSettingEvent(getString(R.string.change_swipe_direction),
                            swipeDirectionStringItems[Integer.valueOf(prefSwipeDirection.value)])
                }
                getString(R.string.key_article_sort) -> {
                    val isNewArticleTop = prefArticleSort.isChecked
                    presenter.updateArticleSort(isNewArticleTop)
                    // GA
                    TrackerHelper.sendSettingEvent(getString(R.string.change_aricle_sort),
                            if (prefArticleSort.isChecked) getString(R.string.article_sort) else getString(R.string.not_article_sort))
                }
                getString(R.string.key_internal_browser) -> {
                    val isInternal = prefInternalBrowser.isChecked
                    presenter.updateInternalBrowser(isInternal)
                    // GA
                    TrackerHelper.sendSettingEvent(getString(R.string.change_browser_option),
                            if (prefInternalBrowser.isChecked) getString(R.string.open_internal) else getString(R.string.not_open_internal))
                }
                getString(R.string.key_auto_update_in_main_ui) -> {
                    val isAutoUpdateInMainUi = prefAutoUpdateInMainUi.isChecked
                    presenter.updateAutoUpdateInMainUi(isAutoUpdateInMainUi)
                    TrackerHelper.sendSettingEvent(getString(R.string.change_auto_update_in_main_ui_option),
                            if (isAutoUpdateInMainUi) getString(R.string.auto_update_in_main_ui) else getString(R.string.not_auto_update_in_main_ui))
                }
                getString(R.string.key_launch_tab) -> {
                    val launchTab = Integer.valueOf(prefLaunchTab.value)
                    presenter.updateLaunchTab(launchTab)
                }
            }
        }

        if (BuildConfig.DEBUG) {
            val prefImport = findPreference<Preference>(getString(R.string.key_import_db))
            activity?.let { activity ->
                prefImport.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    launch {
                        val currentDb = activity.getDatabasePath(DatabaseHelper.DATABASE_NAME)
                        presenter.onImportDatabaseClicked(currentDb)
                        ToastHelper.showToast(activity, getString(R.string.import_db), Toast.LENGTH_SHORT)
                    }
                    true
                }
                val prefExport = findPreference<Preference>(getString(R.string.key_export_db))
                prefExport.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    launch {
                        val currentDb = activity.getDatabasePath(DatabaseHelper.DATABASE_NAME)
                        presenter.onExportDatabaseClicked(currentDb)
                        ToastHelper.showToast(activity, getString(R.string.export_db), Toast.LENGTH_SHORT)
                    }
                    true
                }
                val prefAddRss = findPreference<Preference>(getString(R.string.key_add_rss))
                prefAddRss.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                    launch {
                        presenter.onDebugAddRssClicked()
                    }
                    true
                }
            }
        }

        prefLicense.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            presenter.onLicenseClicked()
            true
        }

        prefReview.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            TrackerHelper.sendButtonEvent(getString(R.string.tap_go_to_google_play_from_setting))
            activity?.let { a ->
                try {
                    val uri = Uri.parse("market://details?id=${a.packageName}")
                    a.startActivity(Intent(Intent.ACTION_VIEW, uri))
                } catch (e: Exception) {
                }
            }
            true
        }

        prefRequest.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            TrackerHelper.sendButtonEvent(getString(R.string.tap_request_from_setting))
            activity?.startActivity(Intent(context, UserRequestActivity::class.java))
            true
        }
    }

    override fun setUpdateInterval(index: Int, summary: String) {
        prefUpdateInterval.setValueIndex(index)
        prefUpdateInterval.summary = summary
    }

    override fun setAutoUpdateInMainUi(isAutoUpdateInMainUi: Boolean) {
        prefAutoUpdateInMainUi.isChecked = isAutoUpdateInMainUi
    }

    override fun setArticleSort(isNewArticleTop: Boolean) {
        prefArticleSort.isChecked = isNewArticleTop
    }

    override fun setInternalBrowser(isEnabled: Boolean) {
        prefInternalBrowser.isChecked = isEnabled
    }

    override fun setAllReadBehavior(index: Int, summary: String) {
        prefAllReadBehavior.setValueIndex(index)
        prefAllReadBehavior.summary = summary
    }

    override fun setSwipeDirection(index: Int, summary: String) {
        prefSwipeDirection.setValueIndex(index)
        prefSwipeDirection.summary = summary
    }

    override fun setLaunchTab(index: Int, summary: String) {
        prefLaunchTab.setValueIndex(index)
        prefLaunchTab.summary = summary
    }

    override fun startLicenseActivity() {
        startActivity(Intent(activity, LicenseActivity::class.java))
    }
}
